### PR TITLE
Replace HashHelpers.Combine with HashCode.Combine

### LINKED ISSUE #25986: Replace HashHelpers.Combine with HashCode.Combine
`System.HashCode` was recently added in order to make high-quality fast hashing easy. We should replace usage of `HashHelpers.Combine`, an internal low-quality hash combiner with `HashCode.Combine`. It should lead to better distribution in dictionaries and the hashing should be nearly as fast.

See: https://github.com/dotnet/coreclr/issues/15508

### PR DESCRIPTION
**Pull request by jcdickinson**
Fixes #25986

All instances of HashHelpers.Combine and NumericsHelpers.CombineHash have been replaced. Copied code has also been replaced as far as grep is concerned. The helpers have been deleted.

### CODE REVIEW

File: src/System.Composition.TypedParts/src/System/Composition/TypedParts/Discovery/DiscoveredPart.cs [hunk 1]
```diff
@@ -236,10 +235,10 @@ private sealed class ParameterInfoComparer : IEqualityComparer<ParameterInfo>
         {
             public static readonly ParameterInfoComparer Instance = new ParameterInfoComparer();
 
-            public int GetHashCode(ParameterInfo obj)
-            {
-                return HashHelpers.Combine(obj.Position.GetHashCode(),  obj.Member.GetHashCode());
-            }
+            public int GetHashCode(ParameterInfo obj) => System.HashCode.Combine(
```
Comment (jcdickinson) [inline src/System.Composition.TypedParts/src/System/Composition/TypedParts/Discovery/DiscoveredPart.cs]: No idea how to get this to compile locally, irregardless of what happens with the CI build that's a problem (as it prevents other developers from building locally).

File: src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs [hunk 2]
```diff
@@ -5,7 +5,6 @@
 using System.Collections;
 using System.Collections.Generic;
 using System.Diagnostics;
-using System.Numerics.Hashing;
 using System.Runtime.InteropServices;
 
 namespace System
```
Comment (jcdickinson) [inline src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs]: Does F# depend on the behavior of this, too? (see https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/Tuple.cs#L12)

File: src/System.Numerics.Vectors/src/System/Numerics/Vector.tt [hunk 3]
```diff
@@ -468,49 +467,151 @@ namespace System.Numerics
         /// <returns>The hash code.</returns>
         public override int GetHashCode()
         {
-            int hash = 0;
-
             if (Vector.IsHardwareAccelerated)
```
Comment (jcdickinson) [inline src/System.Numerics.Vectors/src/System/Numerics/Vector.tt]: Is this specialized code a problem?

File: src/System.Drawing.Primitives/src/System/Drawing/Point.cs [hunk 4]
```diff
@@ -179,7 +178,10 @@ public int Y
         ///       Returns a hash code.
         ///    </para>
         /// </summary>
-        public override int GetHashCode() => HashHelpers.Combine(X, Y);
+        public override int GetHashCode() => HashCode.Combine(
+            X,
+            Y
+        );
```
Comment (karelz) [inline src/System.Drawing.Primitives/src/System/Drawing/Point.cs]: why new lines between args? I would make it single line

File: src/System.Numerics.Vectors/src/System/Numerics/Vector.tt [hunk 5]
```diff
@@ -468,49 +467,151 @@ namespace System.Numerics
...
-<#
-            for (int g = 0; g < GetNumFields(type, totalSize); g++)
-            {
-#>
-                    hash = HashHelpers.Combine(hash, this.<#=GetRegisterFieldName(type, g)#>.GetHashCode());
-<#
-            }
-#>
-                    return hash;
+                if (typeof(T) == typeof(Byte))
```
Comment (karelz) [inline src/System.Numerics.Vectors/src/System/Numerics/Vector.tt]: Why does it need to be expanded? Why not just modify the previous for-loop?

File: src/System.Security.Cryptography.Algorithms/tests/RSATests.cs [hunk 6]
```diff
@@ -192,6 +193,28 @@ public void VerifyDataStream_UsesHashDataAndVerifyHash()
             Assert.True(rsa.VerifyData(new MemoryStream(new byte[] { 42 }), new byte[1] { 24 }, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1));
         }
 
+        [Fact]
+        public void RSAEncryptionPadding_Equality()
+        {
```
Comment (karelz) [inline src/System.Security.Cryptography.Algorithms/tests/RSATests.cs]: Why are we adding this test in this PR?

Comment (jcdickinson) [reply src/System.Security.Cryptography.Algorithms/tests/RSATests.cs]: RSAEncryptionPadding equality is currently untested and is changed in the PR.

File: src/System.Composition.TypedParts/tests/DiscoveredPartTests.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+using System.Collections.Generic;
+using Xunit;
+
+namespace System.Composition.TypedParts.Tests
+{
+    public static class DiscoveredPartTests
+    {
+        [Fact]
+        public static void ParameterInfoComparer_GetHashCode()
+        {
```
[the remark this answers is not in the document]
Comment (jcdickinson) [reply src/System.Composition.TypedParts/tests/DiscoveredPartTests.cs]: Same as RSAEncryptionPadding.

File: src/System.Numerics.Vectors/src/System/Numerics/Vector.tt [hunk 8]
```diff
@@ -468,49 +467,151 @@ namespace System.Numerics
...
+                    hashCode.Add(this.register.sbyte_11);
+                    hashCode.Add(this.register.sbyte_12);
+                    hashCode.Add(this.register.sbyte_13);
+                    hashCode.Add(this.register.sbyte_14);
+                    hashCode.Add(this.register.sbyte_15);
+                    return hashCode.ToHashCode();
+                }
+                else if (typeof(T) == typeof(UInt16))
+                {
+                    return HashCode.Combine(
```
Comment (jcdickinson) [inline src/System.Numerics.Vectors/src/System/Numerics/Vector.tt]: The loop is explicit because of these faster calls (HashCode.Combine is unrolled).

**Comment by karelz** [comment]
@morganbr can you please help push this PR forward?

File: src/System.Security.Cryptography.Algorithms/tests/RSATests.cs [hunk 9]
```diff
@@ -192,6 +193,28 @@ public void VerifyDataStream_UsesHashDataAndVerifyHash()
...
+            var paddingType = typeof(RSAEncryptionPadding);
+            var ctorTypes = new Type[] { typeof(RSAEncryptionPaddingMode), typeof(HashAlgorithmName) };
+            ConstructorInfo paddingCtor = paddingType.GetConstructor(BindingFlags.NonPublic | BindingFlags.Instance, null, ctorTypes, null);
```
Comment (morganbr) [inline src/System.Security.Cryptography.Algorithms/tests/RSATests.cs]: Reflecting over framework internals can fail on UAPAoT, but I don't think you need to for this case. Pkcs1 doesn't have an associated hash algorithm, but Oaep does, so you can work with RSAEncryptionPadding.Pkcs1 and call RSAEncryptionPadding.CreateOaep to pick different HashAlgorithmNames.

File: src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs (hunk 2 above)
Comment (morganbr) [reply src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs]: @jkotas , it looks like you added that comment. Can you please expand on what it's about and what might go wrong if we change it?

File: src/System.Composition.TypedParts/src/System/Composition/TypedParts/Discovery/DiscoveredPart.cs (hunk 1 above)
Comment (morganbr) [reply src/System.Composition.TypedParts/src/System/Composition/TypedParts/Discovery/DiscoveredPart.cs]: It looks like System.Composition.TypedParts only targets netstandard, not netcore and HashCode isn't in netstandard (yet). You probably just need to revert this one for now.

**Comment by karelz** [comment]
@morganbr if you think area owners should review, can you please drive that with them? (here's [area owners list](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/issue-guide.md#areas)). It would be great help to move it forward, thanks!

**Comment by morganbr** [comment]
@safern, @eerhardt, @ViktorHofer, can you please confirm that it's ok to change hash codes on types you own in this PR to use a new algorithm whose outputs are randomized per-process?

File: src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs (hunk 2 above)
Comment (jkotas) [reply src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs]: F# depends on the exact hash code for Tuple only. It is ok to change ValueTuple.

Comment (jkotas) [reply src/System.ValueTuple/src/System/ValueTuple/ValueTuple.cs]: This is CoreFX legacy version of ValueTuple. It does not have access to HashCode. I think you should leave it as is.

File: src/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs [hunk 10]
```diff
@@ -971,11 +971,13 @@ public override int GetHashCode()
             AssertValid();
 
             if (_bits == null)
-                return _sign;
-            int hash = _sign;
+                return HashCode.Combine(_sign);
```
Comment (eerhardt) [inline src/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs]: Is the extra call to `HashCode.Combine` worth it here?  The existing code just returned `_sign`, since that is the value when `_bits` is null.

File: src/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs [hunk 11]
```diff
@@ -971,11 +971,13 @@ public override int GetHashCode()
...
+            
+            var hashCode = new HashCode();
+            hashCode.Add(_sign);
             for (int iv = _bits.Length; --iv >= 0;)
-                hash = NumericsHelpers.CombineHash(hash, unchecked((int)_bits[iv]));
-            return hash;
+                hashCode.Add(unchecked((int)_bits[iv]));
```
Comment (eerhardt) [inline src/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs]: Is the cast still necessary?

File: src/System.Numerics.Vectors/src/System/Numerics/Vector3.cs [hunk 12]
```diff
@@ -43,13 +42,7 @@ public partial struct Vector3 : IEquatable<Vector3>, IFormattable
...
         /// </summary>
         /// <returns>The hash code.</returns>
-        public override int GetHashCode()
-        {
-            int hash = this.X.GetHashCode();
-            hash = HashHelpers.Combine(hash, this.Y.GetHashCode());
-            hash = HashHelpers.Combine(hash, this.Z.GetHashCode());
-            return hash;
-        }
+        public override int GetHashCode() => HashCode.Combine(X, Y, Z);
```
Comment (eerhardt) [inline src/System.Numerics.Vectors/src/System/Numerics/Vector3.cs]: (nit) Vector2, 3, and 4 all use different styles here.  Vector3 has the Combine(X, Y, Z) all on one line.  2 and 4 split it on multiple lines.  IMO - putting them all on a single line reads easier.

File: src/System.Numerics.Vectors/src/System/Numerics/Vector.cs [hunk 13]
```diff
@@ -1288,89 +1287,25 @@ public bool Equals(Vector<T> other)
...
+                    typeof(T) == typeof(UInt64) ||
+                    typeof(T) == typeof(Int64) ||
+                    typeof(T) == typeof(Single) ||
+                    typeof(T) == typeof(Double))
+                {
+                    var hashCode = new HashCode();
                     for (int g = 0; g < Count; g++)
                     {
-                        hash = HashHelpers.Combine(hash, ((Double)(object)this[g]).GetHashCode());
+                        hashCode.Add(((T)(object)this[g]).GetHashCode());
```
Comment (eerhardt) [inline src/System.Numerics.Vectors/src/System/Numerics/Vector.cs]: Is the cast still necessary here?  Same goes for the call to `GetHashCode`.
Can't this just be 

`hashCode.Add(this[g]);`

**Comment by eerhardt** [comment]
> can you please confirm that it's ok to change hash codes on types you own in this PR to use a new algorithm whose outputs are randomized per-process?

What's our general stance on GetHashCode?  I would think in general we say that you can't rely on hash codes to be the same across processes, versions of the CLR, etc.  The same was stated here:  https://blogs.msdn.microsoft.com/ericlippert/2011/02/28/guidelines-and-rules-for-gethashcode/

> Rule: Consumers of GetHashCode cannot rely upon it being stable over time or across appdomains

Assuming that's still our general stance, then I think changing the hash code algorithm on the Vector and BigInteger types are fine.

**Comment by karelz** [comment]
@morganbr any update on next steps, missing reviews, etc.?

**Comment by ViktorHofer** [comment]
I agree with what Eric is saying. As we are already using randomized hashing in various places I believe this change should be fine.

**Comment by morganbr** [comment]
Sounds reasonable to me. @jcdickinson, once you resolve the merge conflicts and build failure, I think we can merge this.

File: src/System.Numerics.Vectors/src/System/Numerics/Vector.tt (hunk 3 above)
Comment (eerhardt) [reply src/System.Numerics.Vectors/src/System/Numerics/Vector.tt]: In general, yes.  The rest of the .tt code uses the `supportedTypes` list, but here we are duplicating that list.  If we ever wanted to add a new supported type, this would need to be updated as well.

**Comment by jcdickinson** [comment]
Sorry, been a bit of a rock and a hard place for me. I'm looking at this right now.

File: src/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs (hunk 10 above)
Comment (jcdickinson) [reply src/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs]: From @morganbr:

```
            // Provide a way of diffusing bits from something with a limited
            // input hash space. For example, many enums only have a few
            // possible hashes, only using the bottom few bits of the code. Some
            // collections are built on the assumption that hashes are spread
            // over a larger space, so diffusing the bits may help the
            // collection work more efficiently.
```

Comment (eerhardt) [reply src/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs]: I don't think I follow this logic.  `BigInteger`'s value, when `_bits` is null, is just the `int` value.  Are we doing the same change to `Int32`'s GetHashCode()?

---
In reply to: [167421134](https://github.com/dotnet/corefx/pull/25987#discussion_r167421134) [](ancestors = 167421134)

File: src/Common/src/CoreLib/System/Numerics/Vector.cs [hunk 14]
```diff
@@ -2747,7 +2683,7 @@ public string ToString(string format, IFormatProvider formatProvider)
         [MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
         public static Vector<T> operator ~(Vector<T> value)
         {
-            return s_allOnes ^ value;
+            return allOnes ^ value;
```
Comment (eerhardt) [inline src/Common/src/CoreLib/System/Numerics/Vector.cs]: Does this compile?

File: src/Common/src/CoreLib/System/Numerics/Vector.cs [hunk 15]
```diff
@@ -1301,89 +1301,57 @@ public bool Equals(Vector<T> other)
...
-
             if (Vector.IsHardwareAccelerated)
             {
                 if (typeof(T) == typeof(Byte))
                 {
+					var hash = new HashCode();
                     for (int g = 0; g < Count; g++)
                     {
-                        hash = HashHelpers.Combine(hash, ((Byte)(object)this[g]).GetHashCode());
+                        hash.Add((Byte)(object)this[g]);
```
Comment (eerhardt) [inline src/Common/src/CoreLib/System/Numerics/Vector.cs]: Are these casts still necessary?  `HashCode.Add` takes a `T`, so `hash.Add(this[g])` should be fine, right?

File: src/Common/src/CoreLib/System/Numerics/Vector.cs (hunk 14 above)
Comment (jcdickinson) [reply src/Common/src/CoreLib/System/Numerics/Vector.cs]: Probably not, having issues with building on my machine so I'm leaning on your build server.

**Comment by ahsonkhan** [comment]
@jcdickinson, what is the status of this PR? Any updates?

**Comment by jcdickinson** [comment]
I'm getting a whole bunch of errors like this when I build:

```
D:\Dev\corefx\Tools\sign.targets(113,5): error : D:\Dev\corefx\bin/obj/Windows_NT.AnyCPU.Debug/System.Diagnostics.FileVersionInfo/netcoreapp/System.Diagnostics.FileVersionInfo.dll: PE file is already strong-name signed. [D:\Dev\corefx\src\System.Diagnostics.FileVersionInfo\src\System.Diagnostics.FileVersionInfo.csproj]
```

This means I can't test any of these changes locally, any ideas why this is happening? (Running build.bat from a VS2010 Dev prompt).

**Comment by jcdickinson** [comment]
```
D:\j\workspace\windows-TGrou---2a8f9c29\src\Common\src\CoreLib\System\Numerics\Vector.cs(1317,21): error CS0246: The type or namespace name 'HashCode' could not be found (are you missing a using directive or an assembly reference?) [D:\j\workspace\windows-TGrou---2a8f9c29\src\System.Numerics.Vectors\src\System.Numerics.Vectors.csproj]
```

Should I revert the Vector changes?

**Comment by eerhardt** [comment]
> I'm getting a whole bunch of errors like this when I build:

I've hit those before as well, but I'm not sure exactly what is causing it.  Try `git clean -xdf` from the root of the repo, and build again.

**Comment by eerhardt** [comment]
> Should I revert the Vector changes?

That code gets built for `netfx`, which doesn't have the `HashCode` type yet.  So we'd either need to compile `HashCode` in as an internal type into `System.Numerics.Vectors`, or continue using `HashHelpers` when building for `netfx`.

I think my vote is for the first option... but I'm not guaranteeing it will work perfectly.

File: src/System.Numerics.Vectors/tests/GenericVectorTests.cs [hunk 16]
```diff
@@ -444,45 +444,6 @@ static GenericVectorTests()
...
-        public void GetHashCodeInt32() { TestGetHashCode<Int32>(); }
-        [Fact]
-        public void GetHashCodeUInt64() { TestGetHashCode<UInt64>(); }
-        [Fact]
-        public void GetHashCodeInt64() { TestGetHashCode<Int64>(); }
-        [Fact]
-        public void GetHashCodeSingle() { TestGetHashCode<Single>(); }
-        [Fact]
-        public void GetHashCodeDouble() { TestGetHashCode<Double>(); }
-        private void TestGetHashCode<T>() where T : struct
```
Comment (ahsonkhan) [inline src/System.Numerics.Vectors/tests/GenericVectorTests.cs]: We are removing these tests here. Should we instead just correct them based on the using the new HashCode.Combine or is that unnecessary?

File: src/System.Security.Cryptography.Algorithms/tests/RSATests.cs [hunk 17]
```diff
@@ -4,6 +4,7 @@
 
 using System.IO;
 using System.Linq;
+using System.Reflection;
```
Comment (ahsonkhan) [inline src/System.Security.Cryptography.Algorithms/tests/RSATests.cs]: Why do we need this using directive?

Comment (jcdickinson) [reply src/System.Security.Cryptography.Algorithms/tests/RSATests.cs]: Good spot. It is an artifact from previous changes, will remove.

File: src/System.Numerics.Vectors/tests/GenericVectorTests.cs (hunk 16 above)
Comment (jcdickinson) [reply src/System.Numerics.Vectors/tests/GenericVectorTests.cs]: Actually, I might be able to get these to work. The problem is the seed randomization that HashCode performs.

**Comment by jcdickinson** [comment]
`Vector.Count` is returning nonsensical values. The maximum number of fields for `Vector<Byte>` (according to `Register`) is 16 (`byte_15`), however, Count for `Vector<Byte>` is 32. This seems to occur when `IsHardwareAccelerated==false` (not sure if the value is correct when `IsHardwareAccelerated==true`).

This is breaking the tests, as garbage data is being added to the HashCode in GetHashCode(). All the tests current assume that Count is valid and so continue happily with the garbage data.

The most recent commit demonstrates this with a bunch of failing tests.

**Comment by jkotas** [comment]
> The maximum number of fields for Vector<Byte> (according to Register) is 16 (byte_15)

`Vector<T>` is intrinsic type. Its actual size is determined dynamically by the runtime based on the hardware capabilities. The sources do not match what the type looks like at runtime.

File: src/System.Numerics.Vectors/src/Properties/InternalsVisibleTo.cs [hunk 18]
```diff
@@ -0,0 +1,7 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+
```
Comment (jkotas) [inline src/System.Numerics.Vectors/src/Properties/InternalsVisibleTo.cs]: Why is this needed? We try to stay away from InternalsVisibleTo for testing.

Comment (jcdickinson) [reply src/System.Numerics.Vectors/src/Properties/InternalsVisibleTo.cs]: HashCode has to be made internal to avoid namespace clashes in netfx, and has to be used in a strong manner due to the seed randomization.

**Comment by jcdickinson** [comment]
@jkotas I assume you mean for the accelerated case? I am referring to non-accelerated: https://github.com/dotnet/corefx/pull/25987/files#diff-814a2bbcda0635f5c1385bbce0b9fde8R1400

```
                    // if byte
                    var hash = new HashCode();
                    hash.Add(this.register.byte_0);
                    // ...
                    hash.Add(this.register.byte_15);
                    return hash;
```

In this case there are explicitly 16 fields.

File: src/Common/src/CoreLib/System/Numerics/Vector.cs [hunk 19]
```diff
@@ -1429,109 +1397,77 @@ public override int GetHashCode()
             {
                 if (typeof(T) == typeof(Byte))
                 {
-                    hash = HashHelpers.Combine(hash, this.register.byte_0.GetHashCode());
```
Comment (jcdickinson) [inline src/Common/src/CoreLib/System/Numerics/Vector.cs]: NB: there are exactly 16 fields when Vector is non-accelerated.

**Comment by jcdickinson** [comment]
This is really not working.

* netfx is failing for many reasons, including `Interop` not being available.
* I can't get the HashCode tests to work for netcore targets. I could use a second set of eyes on this.

Considering that netfx is determined to never work - maybe leave this until HashCode lands in NetFX?

**Comment by eerhardt** [comment]
If you want to continue down the path of trying to make `HashCode` work on `netfx`, you will need to include:

https://github.com/dotnet/corefx/blob/9edb1251fd40c1cc75e80cd4f0b3659b1de26aa4/src/Common/src/CoreLib/Interop/Windows/BCrypt/Interop.BCryptGenRandom.cs#L29

Which has the `Interop.GetRandomBytes` method on Windows.  You'll probably also need the [Interop.Libraries.cs](https://github.com/dotnet/corefx/blob/103639b6ff5aa6ab6097f70732530e411817f09b/src/Common/src/CoreLib/Interop/Windows/Interop.Libraries.cs#L9) file.

Although, at this point, it may make sense to just use the old `HashHelpers` code on `netfx` until `HashCode` is available.  Up to you.

**Comment by ahsonkhan** [comment]
@jcdickinson, wee you able to make progress based on @eerhardt's suggestion? Please feel free to ask questions if you are still blocked on something.

**Comment by jcdickinson** [comment]
@ahsonkhan I am seriously between a rock and a hard place right now. I think it would be best if I stepped back from the PR and let someone carry my changes forward.

**Comment by eerhardt** [comment]
The issue this fixes is marked `Future`, and I don't see this change as necessary for `2.1`.

So unless someone else wants to pick the PR up, it probably won't make it into `2.1` unless we have a champion to push it through....

**Comment by danmosemsft** [comment]
Based on above, I'm going to close this pending someone picking it up again and continuing. CErtainly it's a good change in principle. Thanks @jcdickinson for what you have done so far.

**Comment by jcdickinson** [comment]
I'd recommend waiting for HashCode to land in netfx - I strongly suspect that all of the changes would have eventually been reverted due to the cross-framework compatibility issues.
